### DISCUSSION TITLE
Typescript support on next.config

### DISCUSSION THREAD

**Discussion by JesseKoldewijn** [vercel/next.js#49019] [Ideas]
### Goals

1. Would be great that since we went to TS first (since v13) to also support typescript based configs so we could also run type-checking on there
2. Improvements can also include ensuring that when new config options are added or havibg typo's in your cobfig this could be catched
3. Possible invalidation of curtain config fields could also be catched in a nice direct and visible manner by the ts server instead of just when running the project either dev or build.

### Background

To be honest, in my eyes this seems like a feature that has popped in many peoples's heads at this point and would love to see this added.

### Proposal

Rough description: 
- ts config file support
- typescript based validation of config values

**Comment by icyJoseph** [comment]
For completeness sake, adding this JS docs like comment:

```js
/** @type {import('next').NextConfig} */
const config = {
  poweredByHeader: false,
  reactStrictMode: true,
  productionBrowserSourceMaps: true,
  swcMinify: true,
  compiler: {
    styledComponents: true,
  },
};

module.exports = config;
```

Gets half of the job done... I reckon one small issue would be that the `next.config.js` has to be passed around, or rather be present when staring the dev or prod servers... unless you do standalone builds, where it is serialized IIRC. So there'll be a need for a runtime TS interpreter. 

> Note: next.config.js is read during next build and serialized into the server.js output file. If the legacy [serverRuntimeConfig or publicRuntimeConfig options](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) are being used, the values will be specific to values at build time.

**Reply by JesseKoldewijn** [reply]
> For completeness sake, adding this JS docs like comment:
> 
> ```js
> /** @type {import('next').NextConfig} */
> const config = {
>   poweredByHeader: false,
>   reactStrictMode: true,
>   productionBrowserSourceMaps: true,
>   swcMinify: true,
>   compiler: {
>     styledComponents: true,
>   },
> };
> 
> module.exports = config;
> ```
> 
> Gets half of the job done... I reckon one small issue would be that the `next.config.js` has to be passed around, or rather be present when staring the dev or prod servers... unless you do standalone builds, where it is serialized IIRC. So there'll be a need for a runtime TS interpreter.
> 
> > Note: next.config.js is read during next build and serialized into the server.js output file. If the legacy [serverRuntimeConfig or publicRuntimeConfig options](https://nextjs.org/docs/api-reference/next.config.js/runtime-configuration) are being used, the values will be specific to values at build time.

Yes I'm aware of the js docs like comment and its functionality. But this wont work in usecases where you have external typescript-based integrations with the next config. In my eyes the "@type import" snippet just seems a bit like a partial solution and not one that would be preferred in the long run.

**Reply by icyJoseph** [reply]
> In my eyes the "@type import" snippet just seems a bit like a partial solution and not one that would be preferred in the long run.

That's why it only gets half the job done. 

I guess you can always run you own script that TSC compiles into `next.config.js`, but that's also not good. The root blocker right now is that this file is read at runtime, so it would have to be compiled at runtime as well, it is not impossible but a little trickery would be needed.
